x-pack/plugins/cross_cluster_replication/public/app/components/remote_clusters_provider.js

```diff
@@ -27,10 +27,11 @@ export class RemoteClustersProvider extends PureComponent {
       })
     );
     loadRemoteClusters()
+      .then(sortClusterByName)
       .then((remoteClusters) => {
         this.setState({
           isLoading: false,
-          remoteClusters: sortClusterByName(remoteClusters)
+          remoteClusters
```

Comment:
This is directly related to the video you shared to me: "first-class functions". The same way we don't need to be explicit (and use imperative programming) with `map`, and `forEach`. I must say that I'm a bit confused and don't understand why here you think being declarative adds complexity:

```js
loadRemoteClusters()
  .then(sortClusterByName) 
  .then((remoteClusters) => {
    ...
  })
```

but here no

```js
await Promise.all(names.map(deleteCluster));
```

Reply:
Sure, I can explain. I'm measuring complexity by how many "moving parts" are involved in an expression. In the example of piping through `.then` we've created an additional promise, which adds another moving part. By removing the `.then` we've removed that moving part which would be a reduction in complexity according to how I'm measuring it.

In this regard, I think your example of using `Promise.all` is as simple as you can get. It has the minimum number of moving parts required.

In terms of _expressiveness_, `loadRemoteClusters().then(sortClusterByName)` doesn't contain all of the information that `names.map(deleteCluster)` has. The latter example communicate that `deleteCluster` will accept `name` as its argument  because that's what `map` does, but the first example implies that `remoteClusters` will be the argument passed to `sortClusterByName` because we _assume_ that's what `loadRemoteClusters` resolves to.

Anyway, the bottom line is that I don't see the imperative programming style or functional programming style as inherently good or bad, I think they're both useful tools in various situations and I just wanted to share my perspective in this example -- just another way of thinking about code. :)

Reply:
Thanks for explaining your thoughts, I get it. Not saying I share your opinion on complexity but I understand better your idea. I don't have any problem of reverting the change if you prefer it the other way.
This is, in the end, the result of doing so much testing. From time to time you bump on some code and think: "oh, this would be more elegant written this way" (purely at the _preference_ level 😊 )